SmartDeviceLink/private/SDLProtocolReceivedMessageProcessor.m

```diff
@@ -0,0 +1,302 @@
+    UInt32 dataBytesRemaining;
+}
+
+@end
+
+@implementation SDLProtocolReceivedMessageProcessor
+
+-(instancetype)init {
+    self = [super init];
+    if (!self) { return nil; }
+
+    version = 0;
+    encrypted = false;
+    frameType = 0x00;
+    dataLength = 0;
+    dataBytesRemaining = 0; //Counter for the data pump
+
+    // Message management
+    messageDidEnd = 0;
+
+    //Reset state
+    [self resetState];
+    return self;
+}
+
+- (void)resetState{
+    // Flush Buffers
+    self.headerBuffer = [NSMutableData dataWithCapacity:0];
+    self.payloadBuffer = [NSMutableData dataWithCapacity:0];
+    dataBytesRemaining = 0;
+    
+    // Reset state
+    state = START_STATE;
+    prevState = ERROR_STATE;
+}
+
+
+- (void)processReceiveBuffer:(NSData *)receiveBuffer withMessageReadyBlock:(StateMachineMessageReadyBlock)messageReadyBlock{
+    const BytePtr bytes = (BytePtr)receiveBuffer.bytes;
+    
+    for (int i = 0; i < [receiveBuffer length]; i++) {
+        messageDidEnd = [self sdl_processMessagesStateMachine:(Byte)bytes[i]];
+
+        // If we have reached the end of a message, we need to immediately call the message ready block with the completed data, then reset the buffers and keep pumping data into the state machine
+        if (messageDidEnd){
+            messageDidEnd = 0;
+            messageReadyBlock(header.encrypted, header, [NSData dataWithData:self.payloadBuffer]);
+            [self resetState];
+            return;
+        }
+    }
+    
+}
+
+/// This is the state machine. It processes a single byte of a message, checks for errors, and builds up a header buffer and a payload buffer.
+/// When the header and payload are complete, this returns true to notify the calling funciton.
+/// For reference: https://smartdevicelink.com/en/guides/sdl-overview-guides/protocol-spec/
+/// If a byte comes in that does not conform to spec, the buffers are flushed and state is reset.
+/// @param currentByte The byte currently beign processed
+- (BOOL)sdl_processMessagesStateMachine:(Byte)currentByte {
+    SDLServiceType serviceType = 0x00;
+    SDLFrameInfo controlFrameInfo = 0x00;
+    BOOL endOfMessageFlag = NO;
+    
+    switch (state){
+        case START_STATE:
+            //Flush the buffers
+            [self resetState];
+            
+            // 4 bits for version
+            // 4 highest bits (b1111 0000)
+            version = (currentByte & 0xF0 ) >> 4;
+
+            // 1 bit for either encryption or compression, depending on version. 4th lowest bit (b0000 1000)
+            encrypted = (currentByte & 0x08 ) >> 3;
+            
+            // 3 bits for frameType. 3 lowest bits (b0000 0111)
+            frameType = (currentByte & 0x07) >> 0;
+
+            state = SERVICE_TYPE_STATE;
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            
+            // Check version for errors
+            if ((version < 1 || version > 5)) {
+                prevState = state;
```

Comment:
Easier than doing this everywhere is probably to, once you have your private vars as actual properties, set up a custom setter than automatically sets `prevState` appropriately.

Although, now that I'm looking at it, it doesn't look like you use `prevState` for anything so maybe you could just remove it?

Reply:
`prevState` is a debugging tool.  If you have a breakpoint in the `ERROR_STATE` then `prevState` will tell you what state caused the error.  Is there a better way to do this?

Reply:
There should be logs when you go into the error state. Then you can remove `prevState`. We shouldn't track state for debugging purposes.